File: sdks/typescript/examples/taker-bot.ts
/**
 * O2 Exchange SDK - Taker Bot Example
 *
 * WebSocket-driven taker bot that:
 * - Monitors order book depth in real-time
 * - Executes BoundedMarket orders when price crosses threshold
 * - Configurable target price and slippage tolerance
 *
 * Run: npx tsx examples/taker-bot.ts
 */

import {
  boundedMarketOrder,
  DepthBook,
  formatPrice,
  formatQuantity,
  Network,
  O2Client,
  O2Error,
  scaleQuantityForMarket,
} from "@o2exchange/sdk";

// ── Configuration ─────────────────────────────────────────────────

const CONFIG = {
  network: Network.TESTNET,
  buyBelowPrice: 0.02, // Buy when best ask drops below this
  maxQuantity: 50.0, // Max quantity per trade
  slippagePercent: 0.005, // 0.5% slippage tolerance
};

// ── Main ──────────────────────────────────────────────────────────

async function main() {
  const client = new O2Client({ network: CONFIG.network });

  // Setup
  const wallet = O2Client.generateWallet();
  console.log(`Wallet: ${wallet.b256Address}`);

  const { tradeAccountId } = await client.setupAccount(wallet);
  console.log(`Trade account: ${tradeAccountId}`);

  await sleep(3000);

  // Get market
  const markets = await client.getMarkets();
  const market = markets[0];
  const pair = `${market.base.symbol}/${market.quote.symbol}`;
  console.log(`Monitoring: ${pair}`);

  // Create session (stored on client, tradeAccountId resolved from wallet)
  await client.createSession(wallet, [pair], 30);

  // Stream depth
  console.log(`Watching for asks below ${CONFIG.buyBelowPrice} ${market.quote.symbol}...`);
  const depthStream = await client.streamDepth(pair, 1);

  // Depth messages after the snapshot carry signed relative changes;
  // DepthBook applies them and keeps a correct local book.
  const book = new DepthBook();
  for await (const update of depthStream) {
    book.apply(update);
    const bestAskPrice = book.bestAsk;
    if (bestAskPrice === undefined) continue;

    const bestAsk = formatPrice(market, bestAskPrice);
    const bestAskQty = formatQuantity(market, book.asks.get(bestAskPrice) ?? 0n);

    console.log(
      `Best ask: ${bestAsk.toFixed(6)} ${market.quote.symbol} (qty: ${bestAskQty.toFixed(3)})`,
    );

    // Check if price meets our target
    if (bestAsk <= CONFIG.buyBelowPrice && bestAsk > 0) {
      console.log(`Target price reached! Executing buy...`);

      // Use bigint prices directly from the book (pass-through path)
      const bestAskBigint = bestAskPrice;
      const bestAskQtyBigint = book.asks.get(bestAskPrice) ?? 0n;
      const maxOrderQuantity = scaleQuantityForMarket(market, CONFIG.maxQuantity);

      // Calculate max price with slippage as a string for BoundedMarket
      const maxPrice = (bestAsk * (1 + CONFIG.slippagePercent)).toFixed(6);

      try {
        const response = await client.createOrder(
          pair,
          "buy",
          bestAskBigint, // bigint pass-through — already scaled
          bestAskQtyBigint > maxOrderQuantity ? maxOrderQuantity : bestAskQtyBigint,
          { orderType: boundedMarketOrder(maxPrice, "0") },
        );

        console.log(`Buy executed! TX: ${response.txId}`);

        if (response.orders) {
          for (const order of response.orders) {
            console.log(
              `  Order ${order.order_id}: ${order.side} ${formatQuantity(market, order.quantity)} @ ${formatPrice(market, order.price)}`,
            );
          }
        }
      } catch (error) {
        if (error instanceof O2Error) {
          console.error(`Trade failed: ${error.message}`);
          await client.refreshNonce();
        } else {
          console.error("Unexpected error:", error);
        }
      }
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

main().catch(console.error);

File: sdks/python/examples/taker_bot.py
"""WebSocket-driven taker/sniper bot for the O2 Exchange.

Monitors order book depth via WebSocket and executes BoundedMarket orders
when price crosses a configurable threshold.
"""

import asyncio
import logging
import signal

from o2_sdk import BoundedMarketOrder, Network, O2Client, OrderSide

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(name)s %(message)s")
logger = logging.getLogger("taker_bot")

# Configuration
PRIVATE_KEY = None  # Set to your owner private key hex, or None to generate
MARKET_PAIR = "FUEL/USDC"
BUY_BELOW_PRICE = 0.020  # Buy when best ask drops below this price
MAX_QUANTITY = 50.0  # Maximum quantity per trade
SLIPPAGE_PCT = 0.005  # 0.5% slippage tolerance


async def main():
    client = O2Client(network=Network.TESTNET)
    shutdown_event = asyncio.Event()

    loop = asyncio.get_event_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, lambda: shutdown_event.set())

    # Setup
    if PRIVATE_KEY:
        owner = client.load_wallet(PRIVATE_KEY)
    else:
        owner = client.generate_wallet()
        logger.info("Generated wallet: %s", owner.b256_address)

    account = await client.setup_account(owner)
    logger.info("Trade account: %s", account.trade_account_id)

    market = await client.get_market(MARKET_PAIR)
    logger.info("Monitoring: %s", market.pair)

    session = await client.create_session(owner=owner, markets=[market.pair], expiry_days=7)

    # Wait for funds
    await asyncio.sleep(5)

    logger.info("Watching for asks below %.6f...", BUY_BELOW_PRICE)

    try:
        async for update in client.stream_depth(market.pair, precision=1):
            if shutdown_event.is_set():
                break

            best_ask = update.changes.best_ask
            if not best_ask:
                continue

            ask_price = market.format_price(int(best_ask.price))
            logger.debug("Best ask: %.6f", ask_price)

            if ask_price <= BUY_BELOW_PRICE:
                logger.info("Price %.6f <= %.6f, executing buy!", ask_price, BUY_BELOW_PRICE)

                max_price = ask_price * (1 + SLIPPAGE_PCT)
                quantity = min(
                    MAX_QUANTITY,
                    market.format_quantity(int(best_ask.quantity)),
                )

                try:
                    result = await client.create_order(
                        session=session,
                        market=market.pair,
                        side=OrderSide.BUY,
                        price=ask_price,
                        quantity=quantity,
                        order_type=BoundedMarketOrder(
                            max_price=max_price,
                            min_price=0.0,
                        ),
                        settle_first=True,
                        collect_orders=True,
                    )

                    if result.success:
                        logger.info("Buy executed! tx: %s", result.tx_id)
                        if result.orders:
                            for order in result.orders:
                                logger.info(
                                    "  %s %s @ %s qty=%s",
                                    order.side,
                                    market.pair,
                                    order.price,
                                    order.quantity,
                                )
                    else:
                        logger.error("Buy failed: %s", result.message)

                except Exception as e:
                    logger.error("Trade error: %s", e)
                    await client.refresh_nonce(session)

    finally:
        await client.close()
        logger.info("Taker bot stopped")


if __name__ == "__main__":
    asyncio.run(main())
